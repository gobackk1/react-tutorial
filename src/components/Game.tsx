import React from "react";
import { Mark } from "../types";
import Board from "./Board";
import { jumpToPast, clickSquare } from "../actions";
import { connect } from "react-redux";
import { AppState } from "../reducers";
import { GameState } from "../reducers/game";

type GameProps = {
  clickSquare: (index: number) => void;
  jumpToPast: (step: number) => void;
  game: GameState;
};
export class Game extends React.Component<GameProps, {}> {
  static defaultProps: any = {};
  render() {
    console.log(this.props);

    const history = this.props.game.history;
    const current = history[this.props.game.stepNumber];
    const winner = calculateWinner(current.squares);
    const status = winner
      ? `winner: ${winner}`
      : `Next player: ${this.props.game.xIsNext ? "X" : "O"}`;

    const moves = history.map((step, move) => {
      const desc = move ? `Go to move #${move}` : "Go to game start";

      return (
        <li key={move}>
          <button onClick={() => this.props.jumpToPast(move)}>{desc}</button>
        </li>
      );
    });

    return (
      <div className="game">
        <div className="game-board">
          <Board
            squares={current.squares}
            onClick={index => this.props.clickSquare(index)}
          />
        </div>
        <div className="game-info">
          <div>{status}</div>
          <ol>{moves}</ol>
        </div>
      </div>
    );
  }
}
const mapStateToProps = (state: AppState) => ({ game: state.game });
const mapDispatchToProps = { clickSquare, jumpToPast };

export default connect(mapStateToProps, mapDispatchToProps)(Game);

const calculateWinner = (squares: Mark[]) => {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6]
  ];
  for (let i = 0; i < lines.length; i++) {
    const [a, b, c] = lines[i];
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
  }
  return null;
};
