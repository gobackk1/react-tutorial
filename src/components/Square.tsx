import React from "react";
import { Mark } from "../types";

type SquareProps = {
  value: Mark;
  onClick: () => void;
};

export const Square: React.FC<SquareProps> = (props: SquareProps) => {
  return (
    <button className="square" onClick={props.onClick}>
      {props.value}
    </button>
  );
};
