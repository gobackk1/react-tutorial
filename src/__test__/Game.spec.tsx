import React from "react";
import { shallow, mount, ReactWrapper } from "enzyme";
import Game from "../components/Game";
import Board from "../components/Board";
import { Square } from "../components/Square";
import { mountToJson } from "enzyme-to-json";

describe("Gameコンポーネントのテスト", () => {
  let wrapper: ReactWrapper;

  beforeEach(() => {
    wrapper = mount(<Game></Game>);
  });

  test("子コンポーネントが存在すること", () => {
    expect(wrapper.find(Board).length).toBe(1);
    expect(wrapper.find(Square).length).toBe(9);
  });

  describe("振る舞いのテスト", () => {
    test("Next Playerが表示されていること", () => {
      const gameInfo = wrapper.find(".game-info div");
      expect(gameInfo.text()).toContain("Next player: X");
      wrapper.find('[data-index="0"]').simulate("click");
      expect(gameInfo.text()).toContain("Next player: O");
    });

    test("勝利者が表示されていること", () => {
      wrapper.setState({
        history: [
          {
            squares: ["X", "X", "X", "O", "O", null, null, null, null]
          }
        ]
      });
      expect(wrapper.find(".game-info div").text()).toContain("winner: X");
    });

    test("履歴が表示されること", () => {
      wrapper.find('[data-index="0"]').simulate("click");
      wrapper.find('[data-index="1"]').simulate("click");
      expect(wrapper.find(".game-info ol li").length).toBe(3);
      expect(wrapper.find(".game-info ol li:first-child").text()).toContain(
        "Go to game start"
      );
      expect(wrapper.find(".game-info ol li:last-child").text()).toContain(
        "Go to move #2"
      );
    });
  });

  test("snapshot", () => {
    const snapshot = mountToJson(wrapper);
    expect(snapshot).toMatchSnapshot();
  });
});
