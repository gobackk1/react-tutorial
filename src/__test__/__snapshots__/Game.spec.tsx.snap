// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gameコンポーネントのテスト snapshot 1`] = `
<Game>
  <div
    className="game"
  >
    <div
      className="game-board"
    >
      <Board
        onClick={[Function]}
        squares={
          Array [
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
      >
        <div>
          <div
            className="board-row"
          >
            <Square
              data-index="0"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
            <Square
              data-index="1"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
            <Square
              data-index="2"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
          </div>
          <div
            className="board-row"
          >
            <Square
              data-index="3"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
            <Square
              data-index="4"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
            <Square
              data-index="5"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
          </div>
          <div
            className="board-row"
          >
            <Square
              data-index="6"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
            <Square
              data-index="7"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
            <Square
              data-index="8"
              onClick={[Function]}
              value={null}
            >
              <button
                className="square"
                onClick={[Function]}
              />
            </Square>
          </div>
        </div>
      </Board>
    </div>
    <div
      className="game-info"
    >
      <div>
        Next player: X
      </div>
      <ol>
        <li
          key="0"
        >
          <button
            onClick={[Function]}
          >
            Go to game start
          </button>
        </li>
      </ol>
    </div>
  </div>
</Game>
`;
